### ISSUE TITLE
Etag generation on getServerSideProps doesn't work with React 18

### ISSUE THREAD

**Issue by thodubois** [vercel/next.js#44805]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64
    Binaries:
      Node: 16.16.0
      npm: 8.11.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.1.2-canary.5
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

Data fetching (gS(S)P, getInitialProps)

### Link to the code that reproduces this issue

https://github.com/thodubois/repro-nextjs-etag-generation

### To Reproduce

- build and start server (npm run build && npm run start)
- navigate to `http://localhost:3000/`
- check response headers

### Describe the Bug

No ETag header on SSR (getServerSideProps) page

Next 13 with React 18
<img width="471" alt="Screenshot 2023-01-12 at 10 15 32" src="https://user-images.githubusercontent.com/37809039/212028845-0d0dd330-4d34-41b5-a9f4-ae0c17950947.png">

Next 12 with React 18
<img width="473" alt="Screenshot 2023-01-12 at 10 18 08" src="https://user-images.githubusercontent.com/37809039/212029117-31ea5257-ba6f-4f7b-9641-1b5cc8de0c31.png">

Next 12 with React 17
<img width="470" alt="Screenshot 2023-01-12 at 10 16 58" src="https://user-images.githubusercontent.com/37809039/212029149-9f1fcf5f-00c4-4acb-a4db-d0e6cfaac54f.png">

### Expected Behavior

ETag header is sent on each request.

**Comment by thodubois** [comment]
Ok, I think I found where this issue come from.
On Next 12.x we have this line :
https://github.com/vercel/next.js/blob/e2c5952ef2fa03d4fb6dc4cb7ff50d26c76e7ea8/packages/next/server/base-server.ts#L1112
In `packages/next/bin/next.ts` it was set to `true`
https://github.com/vercel/next.js/blob/e2c5952ef2fa03d4fb6dc4cb7ff50d26c76e7ea8/packages/next/bin/next.ts#L101
 in function of react version : 
https://github.com/vercel/next.js/blob/e2c5952ef2fa03d4fb6dc4cb7ff50d26c76e7ea8/packages/next/server/utils.ts#L26

but it was remove on Next 13.x : 
https://github.com/vercel/next.js/commit/6028a7a69f6796f9d472d4fe67cefc4843efa09a

All this turns this options to true : 
https://github.com/vercel/next.js/blob/471e80f27b9f7840926c2243952cd3000caa87f3/packages/next/src/server/base-server.ts#L1174

And finally on render, it turns `generateStaticHTML` to false
https://github.com/vercel/next.js/blob/471e80f27b9f7840926c2243952cd3000caa87f3/packages/next/src/server/render.tsx#L1117

and create a `RenderResult` class with a Stream 
https://github.com/vercel/next.js/blob/471e80f27b9f7840926c2243952cd3000caa87f3/packages/next/src/server/render.tsx#L1484

instead of a string :
https://github.com/vercel/next.js/blob/471e80f27b9f7840926c2243952cd3000caa87f3/packages/next/src/server/render.tsx#L1481

I don't know what to do here. I would like to help but I don't know if it's wanted or if its really a bug. If someone can give some insight, it would be great.

**Comment by thodubois** [comment]
@huozhi As you touch this part, do you know if it's normal to not generate Etag on "Dynamic" result ? 
I think the main issue come from this part :
https://github.com/vercel/next.js/blob/471e80f27b9f7840926c2243952cd3000caa87f3/packages/next/src/server/send-payload/index.ts#L65
```
  const payload = result.isDynamic() ? null : await result.toUnchunkedString()

  if (payload) {
    const etag = generateEtags ? generateETag(payload) : undefined
    if (sendEtagResponse(req, res, etag)) {
      return
    }
  }
 ```
 But I'm a bit lost.

**Comment by bobaaaaa** [comment]
Can confirm that no ETags are generated in next 12.3.1 with react 18

**Comment by huozhi** [comment]
We enabled streaming if users are using react 18.x in next 12 for pages. For streaming dynamic response we can't calculate out the etag based on full content, since it might keep streaming out new content later. In the latest canary we've disabled the streaming for pages (related change #45895). so you'll still get the etag. I tried latest canary with react 18 it generated etag for the response.

Will close this as resolved
